lib/inspec/file_provider.rb

```diff
@@ -146,50 +146,56 @@ def read_from_zip(file)
       end
       res
     end
-  end
+  end # class ZipProvider
 
   class TarProvider < FileProvider
     attr_reader :files
 
     def initialize(path)
       @path = path
       @contents = {}
-      @files = []
-      walk_tar(@path) do |tar|
-        @files = tar.find_all(&:file?)
 
-        # delete all entries with no name
-        @files = @files.find_all { |x| !x.full_name.empty? && x.full_name.squeeze("/") !~ %r{\.{2}(?:/|\z)} }
+      here = Pathname.new(".")
 
-        # delete all entries that have a PaxHeader
-        @files = @files.delete_if { |x| x.full_name.include?("PaxHeader/") }
+      walk_tar(@path) do |entries|
+        entries.each do |entry|
+          name = entry.full_name
 
-        # replace all items of the array simply with the relative filename of the file
-        @files.map! { |x| Pathname.new(x.full_name).relative_path_from(Pathname.new(".")).to_s }
+          # rubocop:disable Layout/MultilineOperationIndentation
+          # rubocop:disable Style/ParenthesesAroundCondition
+          next unless (entry.file? &&              # duh
+                       !name.empty? &&             # for empty filenames?
+                       name !~ %r{\.\.(?:/|\z)} && # .. (to avoid attacks?)
+                       !name.include?("PaxHeader/"))
+
+          path = Pathname.new(name).relative_path_from(here).to_s
+
+          @contents[path] = begin # not ||= in a tarball, last one wins
```

Comment:
Maybe this begin/end block could be moved out to a separate method like `attempt_utf8_encoding(entry.read)`

Just to avoid a hanging begin/end. But how it is currently seems good to merge too.